pkg/module/blob/blob.go

```diff
@@ -0,0 +1,315 @@
+				ExcludeFiles:   input.ExcludeFiles,
+				PreserveDir:    input.PreserveDir,
+				FollowSymlinks: input.FollowSymlinks,
+			}); err != nil {
+				return nil, fmt.Errorf("unable to tar input artifact: %w", err)
+			}
+			if err := gw.Close(); err != nil {
+				return nil, fmt.Errorf("unable to close gzip writer: %w", err)
+			}
+		} else {
+			input.SetMediaTypeIfNotDefined(MediaTypeTar)
+			if err := TarFileSystem(ctx, fs, input.Path, &data, TarFileSystemOptions{
+				IncludeFiles:   input.IncludeFiles,
+				ExcludeFiles:   input.ExcludeFiles,
+				PreserveDir:    input.PreserveDir,
+				FollowSymlinks: input.FollowSymlinks,
+			}); err != nil {
+				return nil, fmt.Errorf("unable to tar input artifact: %w", err)
+			}
+		}
+
+		return &Output{
+			Digest: digest.FromBytes(data.Bytes()).String(),
+			Size:   int64(data.Len()),
+			Reader: ioutil.NopCloser(&data),
+		}, nil
+	} else if input.Type == FileInputType {
+		if inputInfo.IsDir() {
+			return nil, fmt.Errorf("resource type is file but a directory was provided")
+		}
+		// otherwise just open the file
+		inputBlob, err := fs.Open(input.Path)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read input blob from %q: %w", input.Path, err)
+		}
+		blobDigest, err := digest.FromReader(inputBlob)
+		if err != nil {
+			return nil, fmt.Errorf("unable to calculate digest for input blob from %q, %w", input.Path, err)
+		}
+		if _, err := inputBlob.Seek(0, io.SeekStart); err != nil {
+			return nil, fmt.Errorf("unable to reset input file: %s", err)
+		}
+
+		if input.Compress() {
+			input.SetMediaTypeIfNotDefined(MediaTypeGZip)
+			var data bytes.Buffer
+			gw := gzip.NewWriter(&data)
+			if _, err := io.Copy(gw, inputBlob); err != nil {
+				return nil, fmt.Errorf("unable to compress input file %q: %w", input.Path, err)
+			}
+			if err := gw.Close(); err != nil {
+				return nil, fmt.Errorf("unable to close gzip writer: %w", err)
+			}
+
+			return &Output{
+				Digest: digest.FromBytes(data.Bytes()).String(),
+				Size:   int64(data.Len()),
+				Reader: ioutil.NopCloser(&data),
+			}, nil
+		}
+		return &Output{
+			Digest: blobDigest.String(),
+			Size:   inputInfo.Size(),
+			Reader: inputBlob,
+		}, nil
+	} else {
+		return nil, fmt.Errorf("unknown input type %q", input.Path)
+	}
+}
+
+// TarFileSystemOptions describes additional options for tarring a filesystem.
+type TarFileSystemOptions struct {
+	IncludeFiles []string
+	ExcludeFiles []string
+	// PreserveDir defines that the directory specified in the Path field should be included in the blob.
+	// Only supported for Type dir.
+	PreserveDir    bool
+	FollowSymlinks bool
+
+	root string
+}
+
+// Included determines whether a file should be included.
+func (opts *TarFileSystemOptions) Included(path string) (bool, error) {
+	// if a root path is given remove it rom the path to be checked
+	if len(opts.root) != 0 {
+		path = strings.TrimPrefix(path, opts.root)
+	}
+	// first check if a exclude regex matches
```

Comment:
```suggestion
	// First, check if an exclude regex matches
```
Or is it the "first check"? Then drop my comma :D